### PR TITLE
LPS-195155 Implement saving of FDS item actions in DSM

### PR DESCRIPTION
**Pull request by markocikos**
### References

- [LPS-192149 Create link item actions](https://issues.liferay.com/browse/LPS-192149)

### What is the goal of this PR?

We are implementing persistence for FDS Actions, UI to save and reorder FDS Actions, and doing some minor improvements on the way.

See technical notes inline.

TODO in a separate task:
- we could have icon instead of icon name in Icon field in the orderable table.
- we could have localized Type in the orderable table.

### What does it look like?

https://github.com/liferay-frontend/liferay-portal/assets/5435511/f89ed31c-e4a4-4590-8edf-cb0744eeac1c

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Sorting.tsx [hunk 1]
```diff
@@ -401,15 +399,12 @@ const Sorting = ({
...
 
-				if (storedFDSSorts.length > fdsSortsOrderArray.length) {
-					notOrdered = storedFDSSorts.filter(
-						(filter) =>
-							!fdsSortsOrderArray.includes(String(filter.id))
-					);
-				}
+				notOrdered = storedFDSSorts.filter(
+					(filter) => !fdsSortsOrderArray.includes(String(filter.id))
+				);
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Sorting.tsx:404]: This is a bugfix of a case I found by accident. It is possible that the ordered CSV contains a lot of values, and a smaller amount of saved value(s) are not in it. This happens when you delete items externally. 

Steps:
1. Create few actions in DSM UI. 
2. Save in any order.
3. Delete all action items through published "FDS Action" Objects page.
4. Create an item in DSM UI.

What happens:
- New item is not displayed. Length check in this line does not include it.

I applied all ordering fixes to Sorts, Filters and Actions.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Sorting.tsx [hunk 2]
```diff
@@ -401,15 +399,12 @@ const Sorting = ({
...
 			}
 
-			setFDSSorts([...ordered, ...notOrdered]);
+			setFDSSorts([...notOrdered, ...ordered]);
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Sorting.tsx:407]: Place unordered before ordered, to ensure new items are on top.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Sorting.tsx [hunk 3]
```diff
@@ -376,14 +376,12 @@ const Sorting = ({
 	useEffect(() => {
 		const getFDSSort = async () => {
 			const response = await fetch(
-				`${API_URL.FDS_VIEWS}/${fdsView.id}?nestedFields=${OBJECT_RELATIONSHIP.FDS_VIEW_FDS_SORT}`
+				`${API_URL.FDS_SORTS}?filter=(${OBJECT_RELATIONSHIP.FDS_VIEW_FDS_SORT_ID} eq '${fdsView.id}')&nestedFields=${OBJECT_RELATIONSHIP.FDS_VIEW_FDS_SORT}&sort=dateCreated:desc`
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Sorting.tsx:379]: Sort by creation date, to put the newest unordered item first.

Use exact resource endpoint, that supports sorting, pagination etc.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Actions.tsx [hunk 4]
```diff
@@ -66,21 +73,266 @@ const TYPES = [
...
+	}, [spritemap]);
+
+	const labelFormElementId = `${namespace}Label`;
+	const iconFormElementId = `${namespace}Icon`;
+	const typeFormElementId = `${namespace}Type`;
+	const urlFormElementId = `${namespace}URL`;
+	const confirmationMessageFormElementId = `${namespace}ConfirmationMessage`;
+	const confirmationMessageTypeFormElementId = `${namespace}ConfirmationMessageType`;
+
+	const ConfirmationMessageRow = () => (
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Actions.tsx:280]: Buying space. No state in this component, so we should be safe,

**Comment by markocikos** [comment]
One dependency merged, one to go.

**Comment by markocikos** [comment]
I added two commits to PR, one applying new toast utils throughout DSM, and another with minor style fixes. I don't plan on adding any additional changes. As soon as the spritemap PR is merged, I'll set this as ready for review.

**Comment by markocikos** [comment]
Force-pushed, no changes in code. All dependencies are merged. Open for reviews!

**Comment by markocikos** [comment]
ci:test:sf

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/RequiredMark.tsx [hunk 5]
```diff
@@ -8,7 +8,7 @@ import React from 'react';
 
 const RequiredMark = () => (
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/RequiredMark.tsx:9]: It seems a good candidate to be moved to a shared place as it appears in 3 different modules with the same markup :wink: (adaptive-media, frontend-data-set & frontend-data-set-view)

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/portlet/FDSViewsPortlet.java [hunk 6]
```diff
@@ -104,6 +104,46 @@ protected void doDispatch(
 		super.doDispatch(renderRequest, renderResponse);
 	}
 
+	private void _addLocalizedCustomObjectField(
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/portlet/FDSViewsPortlet.java:107]: :star_struck:

**Review by juanjofgliferay — APPROVED** [review]
LGTM :clap:

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Actions.tsx (hunk 4 above)
Comment (kresimir-coko) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/Actions.tsx:280]: I spent quite some time debugging this, but abstracting the row into this component bugged the `InputLocalized` component inside it, so I will be removing this abstraction in a follow-up. cc @dsanz this is the thing we talked about on friday

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/RequiredMark.tsx (hunk 5 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/RequiredMark.tsx:9]: See https://github.com/liferay/clay/issues/5083

**Review by kresimir-coko — APPROVED** [review]

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/RequiredMark.tsx (hunk 5 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/RequiredMark.tsx:9]: Great

**Comment by markocikos** [comment]
> I spent quite some time debugging this, but abstracting the row into this component bugged the InputLocalized component inside it, so I will be removing this abstraction in a follow-up. cc @dsanz this is the thing we talked about on friday

@kresimir-coko Hmm that's interesting. I was convinced that we can do this, as long as `ConfirmationMessageRow` does not have inner state. It may be that I am wrong. You can return this inline, but there will be a **lot** of line breakage. Alternative way to fix it should be to move this in root scope, and pass in required props.

**Comment by markocikos** [comment]
ci:forward

**Comment by markocikos** [comment]
ci:forward

**Comment by markocikos** [comment]
ci:forward

**Comment by JerryAdrianoo** [comment]
Hey @markocikos these failures are not related to your changes! 
[Testray Comparison.](https://testray.liferay.com/home/-/testray/runs/compare?compare=true&view=details&statusA=2&testrayRunIdB=2422751191&testrayRunIdA=2422541312&statusB=3)

**Comment by markocikos** [comment]
ci:forward:force

**Comment by markocikos** [comment]
Manual forward https://github.com/brianchandotcom/liferay-portal/pull/140636

**Pull request closed without merging**